### ISSUE TITLE
Fully qualified calls and the foo:bar function hack

### ISSUE THREAD

**Issue by rvirding** [lfe/lfe#317]
TL;DR calling a function in another module using mod:fun is a big hack and should be fixed.

I will describe the problem starting from the bottom and work my way up. The core form for making a fully qualified call is `call`:

```(call mod fun arg-1 arg-2 ...)```

where all the arguments are evaluated. The `mod` and `fun` should result in the module and function names to be called. To make it slightly easier to use and a little clearer there is the `:` macro:

```(: mod fun arg-1 arg-2 ...)```

which just quotes mod and fun. It is a clean macro to use. Note that both of these are safe to use as they make no assumptions about the module or function names as long as they are atoms. If you are generating code which makes fully qualified code then it is these you should generate.

People felt that the `:` macro was a bit ugly to use and preferred something which looked more like conventional Erlang. This is where the hack comes in. So at *macro expansion time* the function name in a call is checked and if it contains one, and only one, colon with characters on both sides of it is expanded to a `call`. So

```(foo:bar 1 2) ===> (call foo bar 1 2)```

Note that as this is done at macro expansion time the only way we can affect this is by using an atom with a colon in it and if there is a colon the expansion is always done. There is no way we can control this so we cannot have module or functions names which contain a colon. (we special case the operator `=:=`) So trying to call `(erlang:=:= a b)` just won't work.

Note that in Common Lisp this is no problem as the `:` separates the package containing the symbol and the symbol name itself. And as this is done at read time it is possible to quote the `:` so it just becomes a character in the symbol.

Note again that neither the `:` macro nor the `call` form make any assumptions on the module or function names so they are always safe to use. Calling `(: erlang =:= a b)` will work.

The question is then whether this hack is acceptable or whether we should try and clean it up, and if so how? Perhaps by handling atoms containing a `:` in similar fashion to CL so `foo:bar` could become `#(: foo bar)` and allow quoting `:` in an atom to stop this. Then `foo\:bar` is just an atom and `foo:bar\:baz` would be legal.

Thoughts?

**Comment by oubiwann** [comment]
As you know about me, I _really_ like having complete freedom when it comes to naming things :-) I think it would be fantastic to be allowed to use `:` in names, providing developers with some interesting possibilities. The fact that this would mean cleaning up the hack that we added a couple of years ago is even better :-)

I think escaping is a good approach, following in CL's footsteps. However, I'd be even happier if we could also handle `foo:bar:baz`, where the first colon would be interpreted as the mod/func separator and any subsequent colons would be escaped under the covers ...

**Comment by saaji** [comment]
My (naive) 2 cents. As far as I can see, Erlang also does not allow to use function names with `:` unquoted. It forces to quote function names with `:`

```
Eshell V8.0  (abort with ^G)
1> erlang:=:=(1, 1).
* 1: syntax error before: ':='
1> erlang:'=:='(1, 1).
true
```

Based on Erlang example, one of possible solutions would be to have some form of function an module name quoting. The direct translation of the above example to LFE atom syntax would look like `(erlang:'|=:=| 1 1)`. I must admit that this is far from pretty, especially if one also wants to apply the same rule to module names: `('|some:module|:'|some:fn| arg1 arg2)`.

What could possibly be done here is to have some sort of syntactic sugar around atoms, like allowing to put them in single quotes, i.e `'my:function'`.  Then above examples would look like `(erlang:'=:=' 1 1)` or even `('some:module':'some:fn' arg1 arg2)`. 

What should be noted here, is that CL-like approach with quoting allows the same effect with `(some\:module:some\:fn arg1 arg2)`

**Comment by rvirding** [comment]
It's a bit different between Erlang and LFE depending on when the handling of the `:` occurs. So in Erlang when I write `foo:bar` I get 3 tokens, the atom `foo`, the `:` token and the atom `bar`. It is then the parser which decides if this is a legitimate inter-module call or a syntax error. In LFE now it is different as writing `foo:bar` will result in 1 atom which won't cause a syntax error. It is later in the macro expansion ***after*** parsing where an atom in the function position is checked if it contains a colon and then transforms it to an inter-module call. At this stage it is irrelevant how the `:` got there so no amount of quoting will change it. If it's there, it's there.

The issue is how to handle the colon in the first place, perhaps make it behave more like a `:` in CL which uses it to separate the package name from the symbol. Seeing we don't have packages in LFE, and can't for that matter, we have to do something else. One idea is that `foo:bar` would parse to the tuple `#(: foo bar)` or perhaps `#(foo : bar)`  which could then be cleanly handled to a remote call, and allow for quoting the `:` in the atom. Just like in CL.

The main problem I see with doing that is that that expansion would ***always*** be done as the parser has no real way of knowing if this is inter-module call or not. Maybe this is not a problem just a bit of new syntax, but it would be strange if you write `(foo 1 bar:baz 2)` which you become `(foo 1 #(: bar baz) 2)`. Perhaps not, I am still thinking about this.

Maybe we should just leave it. Unfortunately oubiwann's suggestion would not allow a colon in the module name.

I hope this clarifies it a bit and does not make it harder to understand.

**Comment by saaji** [comment]
Thank you for detailed insights, that is really helpful. I perfectly understand that current implementation parses `foo:bar` to a single atom and all implications of that. As far as I can see, this is because atoms are always quoted in LFE, hence the need to parse atoms with `:` and interpret them in a special way. 

Interesting, what if there was a way to have typographically limited atoms of Erlang, maybe with some syntax sugar, would that help the LFE parser to see the the atom `foo`, the `:` token and the atom `bar`?
